library/core/src/slice/sort.rs

```diff
@@ -270,11 +313,14 @@ where
     // than two fixed-size arrays of length `BLOCK`. VLAs might be more cache-efficient.
 
     // Returns the number of elements between pointers `l` (inclusive) and `r` (exclusive).
-    fn width<T>(l: *mut T, r: *mut T) -> usize {
+    const fn width<T>(l: *mut T, r: *mut T) -> usize {
         assert!(mem::size_of::<T>() > 0);
         // FIXME: this should *likely* use `offset_from`, but more
         // investigation is needed (including running tests in miri).
-        (r.addr() - l.addr()) / mem::size_of::<T>()
+        // (r.addr() - l.addr()) / mem::size_of::<T>()
+
+        // SAFETY: To be investigated as noted above.
+        (unsafe { r.offset_from(l) as usize }) / mem::size_of::<T>()
```

Comment:
How should the const safety of this be verified?

What should the safety comment be exactly?

cc @RalfJung

Reply:
You need to explain why these pointers are in the same allocation and their distance is an integer multiple of their size.

I am not familiar with this code unfortunately and won't be able to do a detailed review.

Reply:
cc @Gankra since the comment about maybe migrating was by them.